### PR TITLE
[FS] Fix CheckCollateralFullValidationRule

### PR DESCRIPTION
**Pull request by quantumagi**
In comparisons involving `maxReorgLength` and `counterChainConsensusTipHeight` one would expect that the `maxReorgLength` is that of the counter chain. However that is not the case. This PR fixes that by:
- first fixing the `CounterChainNetwork` value on the `CounterChainSettings` class.
- using the value to determine the `maxReorgLength` to use in the comparison.

`CollateralChecker.cs:`
```
        public CollateralChecker(ILoggerFactory loggerFactory, IHttpClientFactory httpClientFactory, ICounterChainSettings settings,
            IFederationManager federationManager, ISignals signals, Network network, IAsyncProvider asyncProvider)
        {
...
            this.maxReorgLength = AddressIndexer.GetMaxReorgOrFallbackMaxReorg(settings.CounterChainNetwork);

...
            lock (this.locker)
            {
                if (heightToCheckAt > this.counterChainConsensusTipHeight - this.maxReorgLength)
                {
                    this.logger.LogTrace("(-)[INVALID_CHECK_HEIGHT]:false");
                    return false;
                }
            }
```

`CheckCollateralFullValidationRule.cs:`
```
        public CheckCollateralFullValidationRule(IInitialBlockDownloadState ibdState, ICollateralChecker collateralChecker,
            ISlotsManager slotsManager, IDateTimeProvider dateTime, Network network, CounterChainNetworkWrapper counterChainNetwork)
        {
            this.network = counterChainNetwork.CounterChainNetwork;
...
        public override Task RunAsync(RuleContext context)
        {
            int maxReorgLength = AddressIndexer.GetMaxReorgOrFallbackMaxReorg(this.network);

            // Check if commitment height is less than `mainchain consensus tip height - MaxReorg`.
            if (commitmentHeight > counterChainHeight - maxReorgLength)
            {
```

### CODE REVIEW

File: src/Stratis.Features.FederatedPeg/Collateral/CheckCollateralFullValidationRule.cs [hunk 1]
```diff
@@ -31,16 +32,16 @@ public class CheckCollateralFullValidationRule : FullValidationConsensusRule
         private readonly int collateralCheckBanDurationSeconds;
 
         public CheckCollateralFullValidationRule(IInitialBlockDownloadState ibdState, ICollateralChecker collateralChecker,
-            ISlotsManager slotsManager, IDateTimeProvider dateTime, Network network)
+            ISlotsManager slotsManager, IDateTimeProvider dateTime, Network network, CounterChainNetworkWrapper counterChainNetwork)
         {
-            this.network = network;
+            this.network = counterChainNetwork.CounterChainNetwork;
```
Comment (fassadlr) [inline src/Stratis.Features.FederatedPeg/Collateral/CheckCollateralFullValidationRule.cs]: `this.network` will need to be renamed as its confusing which one to use down below when calculating `collateralCheckBanDurationSeconds `

File: src/Stratis.Features.FederatedPeg/Collateral/SmartContractCollateralPoARuleRegistration.cs [hunk 2]
```diff
@@ -23,17 +23,20 @@ public class SmartContractCollateralPoARuleRegistration : SmartContractPoARuleRe
...
 
+        private readonly CheckCollateralFullValidationRule checkCollateralFullValidationRule;
+
         private readonly IDateTimeProvider dateTime;
 
         public SmartContractCollateralPoARuleRegistration(Network network, IStateRepositoryRoot stateRepositoryRoot, IContractExecutorFactory executorFactory,
             ICallDataSerializer callDataSerializer, ISenderRetriever senderRetriever, IReceiptRepository receiptRepository, ICoinView coinView,
             IEnumerable<IContractTransactionPartialValidationRule> partialTxValidationRules, IEnumerable<IContractTransactionFullValidationRule> fullTxValidationRules,
-            IInitialBlockDownloadState ibdState, ISlotsManager slotsManager, ICollateralChecker collateralChecker, IDateTimeProvider dateTime)
+            IInitialBlockDownloadState ibdState, ISlotsManager slotsManager, ICollateralChecker collateralChecker, IDateTimeProvider dateTime, CheckCollateralFullValidationRule checkCollateralFullValidationRule)
```
Comment (fassadlr) [inline src/Stratis.Features.FederatedPeg/Collateral/SmartContractCollateralPoARuleRegistration.cs]: Can't we rather inject `CounterchainNetworWorkwrapper` here as opposed to `CheckCollateralFullValidationRule` seeing as we already have it in the DI? Then is saves us from adding to the DI.

Then construct the rule as per usual but with the counter chain network.

File: src/Stratis.Features.FederatedPeg/Collateral/CheckCollateralFullValidationRule.cs (hunk 1 above)
Comment (noescape00) [reply src/Stratis.Features.FederatedPeg/Collateral/CheckCollateralFullValidationRule.cs]: Also on ln71 add a comment that `maxReorgLength` is a counter chain network length.

**Comment by noescape00** [comment]
Overall looks good, just a few readability fixes

**Comment by codingupastorm** [comment]
Agreed, looks good, just the ideas mentioned above. 

We should clean up the `CounterChainSettings / CounterChainNetworkWrapper` mess in the future. And try and only use one and get rid of what's unused.

File: src/Stratis.Features.FederatedPeg/Collateral/SmartContractCollateralPoARuleRegistration.cs (hunk 2 above)
Comment (quantumagi) [reply src/Stratis.Features.FederatedPeg/Collateral/SmartContractCollateralPoARuleRegistration.cs]: Doing it this way requires smart contracts to know less about constructing the collateral rule, which I think is a good thing.

File: src/Stratis.Features.FederatedPeg/Collateral/CheckCollateralFullValidationRule.cs (hunk 1 above)
Comment (quantumagi) [reply src/Stratis.Features.FederatedPeg/Collateral/CheckCollateralFullValidationRule.cs]: Done

**Comment by fassadlr** [comment]
https://dev.azure.com/Stratisplatformuk/StratisBitcoinFullNode/_workitems/edit/4020

**Pull request merged**